nilearn/mass_univariate/_utils.py

```diff
@@ -65,50 +65,68 @@ def _calculate_tfce(
     """
     tfce_4d = np.zeros_like(arr4d)
 
+    # For each passed t map
     for i_regressor in range(arr4d.shape[3]):
         arr3d = arr4d[..., i_regressor]
+
+        # Get signs / threshs
         if two_sided_test:
             signs = [-1, 1]
-
-            # Get the maximum statistic in the map
             max_score = np.max(np.abs(arr3d))
         else:
             signs = [1]
-
-            # Get the maximum statistic in the map
             max_score = np.max(arr3d)
 
         if dh == 'auto':
             step = max_score / 100
         else:
             step = dh
 
-        for score_thresh in np.arange(step, max_score + step, step):
-            for sign in signs:
-                temp_arr3d = arr3d * sign
+        # Set based on determined step size
+        score_threshs = np.arange(step, max_score + step, step)
+
+        # If we apply the sign first...
+        for sign in signs:
+
+            # Then init a temp array, we can re-use the
+            # same array and save a little bit of time
+            temp_arr3d = arr3d.copy() * sign
+
+            # Prep step
+            for score_thresh in score_threshs:
 
-                # Threshold map at *h*
+                # Assumes each score thresh is incrementally larger
                 temp_arr3d[temp_arr3d < score_thresh] = 0
 
-                # Derive clusters
-                labeled_arr3d, n_clusters = label(
-                    temp_arr3d,
-                    bin_struct,
-                )
-
-                # Label each cluster with its extent
-                # Each voxel's cluster extent at threshold *h* is thus *e(h)*
-                cluster_map = np.zeros(temp_arr3d.shape, int)
-                for cluster_val in range(1, n_clusters + 1):
-                    bool_map = labeled_arr3d == cluster_val
-                    cluster_map[bool_map] = np.sum(bool_map)
-
-                # Calculate each voxel's tfce value based on its cluster extent
-                # and z-value
+                # Label into clusters
+                labeled_arr3d, _ = label(temp_arr3d, bin_struct)
+
+                # Get flattened version of the array
+                labeled_arr3d_flat = labeled_arr3d.flatten()
+
+                # And keep track of non-zero inds
+                non_zero_inds = np.where(labeled_arr3d_flat != 0)[0]
+
+                # Get unique values / counts to compute frequencies
+                idx = labeled_arr3d_flat[non_zero_inds]
+                freqs_idx = np.bincount(idx)
```

Comment:
The thing that initially confused me is that `bincount` also includes a bin for zeros, which works out because `idx` starts with one for the first cluster. I think it would be helpful to have a comment here to that effect.

```suggestion
                # bincount includes a bin for zeros, but this bin is skipped 
                # automatically later
                freqs_idx = np.bincount(idx)
```

Reply:
Thanks! I've made your suggestions and added some extra clarity to some other comments also~